src/dataflow/src/render/join.rs

```diff
@@ -46,46 +50,90 @@ where
-
-            // Unwrap demand
-            // TODO: If we pushed predicates into the operator, we could have a
-            // more accurate view of demand that does not include the support of
-            // all predicates.
-            let demand = demand.clone().unwrap_or_else(|| (0..arity).collect());
+            // We maintain a private copy of `map_filter_project`, which we will
+            // digest as we produce the join.
+            let mut mfp = map_filter_project;
 
             // This collection will evolve as we join in more inputs.
+            // TODO(mcsherry): this clashes with the subsequent arrangement-based input.
+            // TODO(mcsherry): determine and apply closure here in `flat_map_ref` form.
+            // TODO(mcsherry): If we plan to use an arrangement, should one exist, then
+            // this is wasteful as it instantiates all rows which are then dropped.
             let (mut joined, mut errs) = self.collection(&inputs[*start]).unwrap();
 
-            // Maintain sources of each in-progress column.
-            let mut source_columns = input_mapper.global_columns(*start).collect::<Vec<_>>();
+            // NOTE(mcsherry): ideally this code is rarely/never relevant, as the associated logic
+            // could be pushed down to the input and perhaps beyond. I'm not certain under what
+            // circumstance we should just delete it, though.
+            let use_leading_arrangement = start_arr.is_some() && inputs.len() > 1;
+            if !use_leading_arrangement {
+                // At this point we are able to construct a per-row closure that can be applied once
+                // we have the first wave of columns in place. We will not apply it quite yet, because
+                // we have three code paths that might produce data and it is complicated.
+                let closure = MyClosure::build(&mut column_map, &mut equivalences, &mut mfp);
 
-            let mut predicates = predicates.to_vec();
-            if start_arr.is_none() || inputs.len() == 1 {
                 // If there is no starting arrangement, then we can run filters
                 // directly on the starting collection.
                 // If there is only one input, we are done joining, so run filters
-                let (j, es) = crate::render::delta_join::build_filter(
-                    joined,
-                    &source_columns,
-                    &mut predicates,
-                    &mut equivalences,
-                );
+                let (j, es) = joined.flat_map_fallible({
+                    // Reuseable allocation for unpacking.
+                    let mut datums = DatumVec::new();
+                    let mut row_packer = RowPacker::new();
+                    move |row| {
+                        let temp_storage = RowArena::new();
+                        let mut datums_local = datums.borrow_with(&row);
+                        // TODO(mcsherry): re-use `row` allocation.
```

Comment:
similarly: i see a lot of `reuse row allocations` sprinkled around these changes - maybe that deserves to be its own isssue

Reply:
I'm ok leaving them as `TODO(mcsherry)` at the moment, but idk. I plan to go through all of this again once this lands, rather than make it more monolithic. I do think we should eventually review all of `render/` for this sort of issue (and a few others that I'll write up).